File: src/idea_enhancer/decision.py
from __future__ import annotations

import logging
import re

from idea_enhancer.client import call
from idea_enhancer.models import DecisionArtifact, Scores
from idea_enhancer.prompts import DECISION_SYSTEM, decision_user

logger = logging.getLogger("idea_enhancer")

SCORE_LINE_RE = re.compile(r"^-\s*([A-Za-z][A-Za-z\s\-/]*?):\s*(\d{1,2})", re.MULTILINE)
RECOMMENDATION_RE = re.compile(
    r"Recommendation:\s*(kill|explore|build)\s*[\.\-:]?\s*(.*)",
    re.IGNORECASE,
)

# Mapping from normalized label → Scores field and human-readable name.
_SCORE_FIELDS: dict[str, tuple[str, str]] = {
    "tam": ("tam", "TAM"),
    "competitive density": ("competitive_density", "Competitive density"),
    "moat": ("moat", "Moat"),
    "founder-fit": ("founder_fit", "Founder-fit"),
    "founder fit": ("founder_fit", "Founder-fit"),
    "why-now timing": ("why_now", "Why-now timing"),
    "why-now": ("why_now", "Why-now timing"),
    "capital efficiency": ("capital_efficiency", "Capital efficiency"),
}


def _section(text: str, header: str) -> str:
    pattern = re.compile(
        rf"^###\s*{re.escape(header)}\s*\n(.*?)(?=^###\s|\Z)",
        re.DOTALL | re.MULTILINE,
    )
    m = pattern.search(text)
    return m.group(1).strip() if m else ""


def _parse_scores(scores_block: str) -> Scores:
    found: dict[str, int] = {}
    for m in SCORE_LINE_RE.finditer(scores_block):
        label = m.group(1).strip().lower()
        try:
            val = max(1, min(10, int(m.group(2))))
        except ValueError:
            continue
        found[label] = val

    # Build Scores with explicit per-field lookup + warning for defaults.
    field_values: dict[str, int] = {}
    for label, (field_name, display_name) in _SCORE_FIELDS.items():
        if label in found:
            field_values[field_name] = found[label]

    # Warn about any dimension that wasn't parsed (will use default 5).
    parsed_fields = set(field_values.keys())
    all_fields = {v[0] for v in _SCORE_FIELDS.values()}
    missing = all_fields - parsed_fields
    if missing:
        missing_names = sorted(
            {v[1] for v in _SCORE_FIELDS.values() if v[0] in missing}
        )
        logger.warning(
            "Decision parser: could not parse scores for %s — defaulting to 5/10. "
            "LLM output may have deviated from expected format.",
            ", ".join(missing_names),
        )

    return Scores(
        tam=field_values.get("tam", 5),
        competitive_density=field_values.get("competitive_density", 5),
        moat=field_values.get("moat", 5),
        founder_fit=field_values.get("founder_fit", 5),
        why_now=field_values.get("why_now", 5),
        capital_efficiency=field_values.get("capital_efficiency", 5),
    )


def _parse_kill_criteria(block: str) -> list[str]:
    out: list[str] = []
    for line in block.splitlines():
        s = line.strip()
        if s.startswith(("-", "*", "•")):
            out.append(s[1:].strip())
    return out[:5]


def _parse_recommendation(block: str) -> tuple[str, str]:
    m = RECOMMENDATION_RE.search(block)
    if not m:
        return "explore", block.strip()
    return m.group(1).lower(), m.group(2).strip().lstrip("—-:.").strip()


def parse_decision(text: str) -> DecisionArtifact:
    scores_block = _section(text, "Scores (1-10, integer)") or _section(text, "Scores")
    kill_block = _section(text, "Kill criteria")
    exp_block = _section(text, "Cheapest validation experiment")
    rec_block = _section(text, "Recommendation")

    scores = _parse_scores(scores_block)
    kill = _parse_kill_criteria(kill_block)
    rec, rationale = _parse_recommendation(rec_block)

    return DecisionArtifact(
        scores=scores,
        kill_criteria=kill or ["(no kill criteria parsed — see raw output)"],
        validation_experiment=exp_block or "(no experiment parsed — see raw output)",
        recommendation=rec,
        rationale=rationale,
    )


def run_decision(
    idea: str, context: str, market: str, competitors: str, synthesis: str
) -> DecisionArtifact:
    raw = call(
        system=DECISION_SYSTEM,
        user=decision_user(idea, context, market, competitors, synthesis),
        max_tokens=2500,
        effort="high",
    )
    return parse_decision(raw)

File: frontend/src/lib/pipeline/decision.ts
/**
 * Decision parser — ported from Python decision.py.
 * Parses the LLM output into a structured DecisionArtifact.
 */

import type { DecisionArtifact, Scores } from '../types';

const SCORE_LINE_RE = /^-\s*([A-Za-z][A-Za-z\s\-/]*?):\s*(\d{1,2})/gm;
const RECOMMENDATION_RE = /Recommendation:\s*(kill|explore|build)\s*[\.\-:]?\s*(.*)/i;

// Mapping from normalized label → Scores field name
const SCORE_FIELDS: Record<string, keyof Scores> = {
  tam: 'tam',
  'competitive density': 'competitive_density',
  'competitive_density': 'competitive_density',
  moat: 'moat',
  'founder-fit': 'founder_fit',
  'founder fit': 'founder_fit',
  'why-now timing': 'why_now',
  'why-now': 'why_now',
  'capital efficiency': 'capital_efficiency',
};

function extractSection(text: string, header: string): string {
  const pattern = new RegExp(
    `^###\\s*${escapeRegex(header)}\\s*\\n([\\s\\S]*?)(?=^###\\s|\\Z)`,
    'm'
  );
  const m = text.match(pattern);
  return m ? m[1].trim() : '';
}

function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function parseScores(scoresBlock: string): Scores {
  const found: Record<string, number> = {};
  let match: RegExpExecArray | null;

  // Reset regex state
  const re = new RegExp(SCORE_LINE_RE.source, 'gm');
  while ((match = re.exec(scoresBlock)) !== null) {
    const label = match[1].trim().toLowerCase();
    const val = Math.max(1, Math.min(10, parseInt(match[2], 10)));
    if (!isNaN(val)) {
      found[label] = val;
    }
  }

  // Build Scores with explicit per-field lookup
  const fieldValues: Partial<Scores> = {};
  for (const [label, fieldName] of Object.entries(SCORE_FIELDS)) {
    if (label in found) {
      (fieldValues as any)[fieldName] = found[label];
    }
  }

  return {
    tam: fieldValues.tam ?? 5,
    competitive_density: fieldValues.competitive_density ?? 5,
    moat: fieldValues.moat ?? 5,
    founder_fit: fieldValues.founder_fit ?? 5,
    why_now: fieldValues.why_now ?? 5,
    capital_efficiency: fieldValues.capital_efficiency ?? 5,
  };
}

function parseKillCriteria(block: string): string[] {
  const out: string[] = [];
  for (const line of block.split('\n')) {
    const s = line.trim();
    if (s.startsWith('-') || s.startsWith('*') || s.startsWith('•')) {
      out.push(s.slice(1).trim());
    }
  }
  return out.slice(0, 5);
}

function parseRecommendation(block: string): [string, string] {
  const m = block.match(RECOMMENDATION_RE);
  if (!m) return ['explore', block.trim()];
  return [m[1].toLowerCase(), m[2].trim().replace(/^[\-—:.]+/, '').trim()];
}

export function parseDecision(text: string): DecisionArtifact {
  const scoresBlock =
    extractSection(text, 'Scores (1-10, integer)') || extractSection(text, 'Scores');
  const killBlock = extractSection(text, 'Kill criteria');
  const expBlock = extractSection(text, 'Cheapest validation experiment');
  const recBlock = extractSection(text, 'Recommendation');

  const scores = parseScores(scoresBlock);
  const kill = parseKillCriteria(killBlock);
  const [rec, rationale] = parseRecommendation(recBlock);

  return {
    scores,
    kill_criteria: kill.length > 0 ? kill : ['(no kill criteria parsed — see raw output)'],
    validation_experiment: expBlock || '(no experiment parsed — see raw output)',
    recommendation: rec as 'kill' | 'explore' | 'build',
    rationale,
  };
}
